lib/billy/ssl/certificate.rb

```diff
@@ -0,0 +1,124 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+require 'openssl'
+require 'fileutils'
+
+module Billy
+  # This class is dedicated to the generation of a request certifcate for a
+  # given domain name. We have to generate for each handled connection a new
+  # request certifcate, due to the fact that each request has propably a
+  # different domain name which will be proxied. So we can't know of future
+  # domain name we could include in the list of subject alternative names
+  # which is required by modern browsers. (Chrome 58+)
+  #
+  # We use our generated certifcate authority to sign any request certifcate,
+  # so a client can be prepared to trust us before a possible test scenario
+  # starts.
+  #
+  # This behaviour and functionality mimics the mighty mitmproxy and it will
+  # enable the usage of Chrome Headless at a time where no ssl issue ignoring
+  # works. And its even secure at testing level.
+  class Certificate
+    attr_reader :key, :cert, :domain
+
+    # To generate a new request certifcate just pass the domain in and you
+    # are ready to go.
+    #
+    # Example:
+    #
+    #   cert = Billy::Certificate.new('localhost')
+    #   [cert.cert_file, cert.key_file]
+    def initialize(domain)
+      @domain = domain
+      @key = OpenSSL::PKey::RSA.new(2048)
+      @cert = generate
+    end
+
+    # Write out the private key to file (PEM format) and give back the
+    # file path. This will produce a temporary file which will be remove
+    # after the current process terminates.
+    def key_file
+      path = File.join(Billy.config.certs_path, "request-#{domain}.key")
+      FileUtils.mkdir_p(File.dirname(path))
+      File.write(path, key.to_pem)
+      path
+    end
+
+    # Write out the certifcate to file (PEM format) and give back the
+    # file path. This will produce a temporary file which will be remove
+    # after the current process terminates.
+    def cert_file
```

Comment:
Can we DRY up the file methods in here too?

Reply:
Yep.